<a href="https://colab.research.google.com/github/luis-fnogueira/exploratory-baking/blob/master/banking_cognitive_class.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Download data source.

In [2]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/00222/bank-additional.zip

--2022-03-09 22:40:43--  https://archive.ics.uci.edu/ml/machine-learning-databases/00222/bank-additional.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 444572 (434K) [application/x-httpd-php]
Saving to: ‘bank-additional.zip’

bank-additional.zip 100%[===================>] 434.15K  1.64MB/s    in 0.3s    

2022-03-09 22:40:43 (1.64 MB/s) - ‘bank-additional.zip’ saved [444572/444572]



Unzipping to a folder. `-o` and `-q` are used to quiet the process and overwrite any existing folders.

In [3]:
!unzip -o -q bank-additional.zip

Importing the libraries necessary for this exercise. Aliases are set to make the libraries easier to use and set a default figure size for further plots. Ignore the warnings.

In [4]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

%matplotlib inline
plt.rcParams["figure.figsize"] = (8, 6)

import warnings
warnings.filterwarnings('ignore')

Here I specify the value of the precision parameter equal to 2 to display two decimal signs (instead of 6 as default).

In [5]:
pd.set_option("precision", 2)
pd.options.display.float_format = '{:.2f}'.format

# Dataset exploration

Reading the 5 first rows of the dataset.

In [6]:
df = pd.read_csv('bank-additional/bank-additional-full.csv', sep = ';')
df.head(5)

age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

  month day_of_week  ...  campaign  pdays  previous     poutcome emp.var.rate  \
0   may         mon  ...         1    999         0  nonexistent         1.10   
1   may         mon  ...         1    999         0  nonexistent         1.10   
2   may         mon  ...         1    999         0  nonexistent         1.10   
3   may         mon  ...         1    999         0  nonexistent         1.10   
4   may         mon  ...         1    999         0  nonexistent         1.10   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0           93.99         -36.40       4.86      5191.00  no  
1           93.99         -36.40       4.86      5191.00  no  
2           93.99         -36.40       4.86      5191.00  no  
3           93.99         -36.40       4.86      5191.00  no  
4           93.99         -36.40       4.86      5191.00  no  

[5 rows x 21 columns]

### Understanding the dataset size, feature names and their types


In [7]:
df.shape

(41188, 21)

The dataset contains 41188 objects (rows), for each of which 21 features are set (columns), including 1 target feature (y).

### Attributing information
Output the column (feature) names:

In [8]:
df.columns

Index(['age', 'job', 'marital', 'education', 'default', 'housing', 'loan',
       'contact', 'month', 'day_of_week', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'emp.var.rate', 'cons.price.idx',
       'cons.conf.idx', 'euribor3m', 'nr.employed', 'y'],
      dtype='object')

Input features (column names):

1.  `age` - client's age in years (numeric)
2.  `job` - type of job (categorical: `admin.`, `blue-collar`, `entrepreneur`, `housemaid`, `management`, `retired`, `self-employed`, `services`, `student`, `technician`, `unemployed`, `unknown`)
3.  `marital` - marital status (categorical: `divorced`, `married`, `single`, `unknown`)
4.  `education` - client's education (categorical: `basic.4y`, `basic.6y`, `basic.9y`, `high.school`, `illiterate`, `professional.course`, `university.degree`, `unknown`)
5.  `default` - has credit in default? (categorical: `no`, `yes`, `unknown`)
6.  `housing` - has housing loan? (categorical: `no`, `yes`, `unknown`)
7.  `loan` - has personal loan? (categorical: `no`, `yes`, `unknown`)
8.  `contact` - contact communication type (categorical: `cellular`, `telephone`)
9.  `month` - last contact month of the year (categorical: `jan`, `feb`, `mar`, ..., `nov`, `dec`)
10. `day_of_week` - last contact day of the week (categorical: `mon`, `tue`, `wed`, `thu`, `fri`)
11. `duration` - last contact duration, in seconds (numeric).
12. `campaign` - number of contacts performed and for this client during this campaign (numeric, includes the last contact)
13. `pdays` - number of days that have passed after the client was last contacted from the previous campaign (numeric; 999 means the client has not been previously contacted)
14. `previous` - number of contacts performed for this client before this campaign (numeric)
15. `poutcome` - outcome of the previous marketing campaign (categorical: `failure`, `nonexistent`, `success`)
16. `emp.var.rate` - employment variation rate, quarterly indicator (numeric)
17. `cons.price.idx` - consumer price index, monthly indicator (numeric)
18. `cons.conf.idx` - consumer confidence index, monthly indicator (numeric)
19. `euribor3m` - euribor 3 month rate, daily indicator (numeric)
20. `nr.employed` - number of employees, quarterly indicator (numeric)

Output feature (desired target):

21. `y` - has the client subscribed a term deposit? (binary: `yes`,`no`)

To see the general information on all the DataFrame features (columns), we use the info method

In [9]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  object 
 2   marital         41188 non-null  object 
 3   education       41188 non-null  object 
 4   default         41188 non-null  object 
 5   housing         41188 non-null  object 
 6   loan            41188 non-null  object 
 7   contact         41188 non-null  object 
 8   month           41188 non-null  object 
 9   day_of_week     41188 non-null  object 
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  object 
 15  emp.var.rate    41188 non-null  float64
 16  cons.price.idx  41188 non-null  float64
 17  cons.conf.idx   41188 non-null 

As you can see, the dataset is full, no pass (`non-null`), so there is no need to fill the gaps. The dataset contains 5 integer (`int64`), 5 real (`float64`) and 11 categorical and binary (`object`) features.

Method **`describe`** shows the main statistical characteristics of the dataset for each numerical feature (`int64` and `float64` types): the existing values number, mean, standard deviation, range, min & max, 0.25, 0.5 and 0.75 quartiles.


In [10]:
df.describe()

age  duration  campaign    pdays  previous  emp.var.rate  \
count 41188.00  41188.00  41188.00 41188.00  41188.00      41188.00   
mean     40.02    258.29      2.57   962.48      0.17          0.08   
std      10.42    259.28      2.77   186.91      0.49          1.57   
min      17.00      0.00      1.00     0.00      0.00         -3.40   
25%      32.00    102.00      1.00   999.00      0.00         -1.80   
50%      38.00    180.00      2.00   999.00      0.00          1.10   
75%      47.00    319.00      3.00   999.00      0.00          1.40   
max      98.00   4918.00     56.00   999.00      7.00          1.40   

       cons.price.idx  cons.conf.idx  euribor3m  nr.employed  
count        41188.00       41188.00   41188.00     41188.00  
mean            93.58         -40.50       3.62      5167.04  
std              0.58           4.63       1.73        72.25  
min             92.20         -50.80       0.63      4963.60  
25%             93.08         -42.70       1.34      5099.10  
50%             93.75         -41.80       4.86      5191.00  
75%             93.99         -36.40       4.96      5228.10  
max             94.77         -26.90       5.04      5228.10

The `Mean` row shows the feature average, `STD` is an RMS (Root Mean Square) deviation, `min`,`  max ` - the minimum and maximum values, `25%`, `50%`, ` 75%  `- quarters that split the dataset (or part of it) into four groups containing approximately an equal number of observations (rows). For example, the duration (`duration`) of about a quarter of calls to customers is around 100 seconds.

The `Mean` row shows the feature average, `STD` is an RMS (Root Mean Square) deviation, `min`,`  max ` - the minimum and maximum values, `25%`, `50%`, ` 75%  `- quarters that split the dataset (or part of it) into four groups containing approximately an equal number of observations (rows). For example, the duration (`duration`) of about a quarter of calls to customers is around 100 seconds.

To see the statistics on non-numeric features, you need to explicitly specify the feature types by the `include` parameter. You can also set `include = all` to output statistics on all the existing features.


In [11]:
df.describe(include = ["object"])

job  marital          education default housing   loan   contact  \
count    41188    41188              41188   41188   41188  41188     41188   
unique      12        4                  8       3       3      3         2   
top     admin.  married  university.degree      no     yes     no  cellular   
freq     10422    24928              12168   32588   21576  33950     26144   

        month day_of_week     poutcome      y  
count   41188       41188        41188  41188  
unique     10           5            3      2  
top       may         thu  nonexistent     no  
freq    13769        8623        35563  36548

The result shows that the average client refers to administrative staff (`job = admin.`), is married (`marital = married`) and has a university degree (`education = university.degree`).

For categorical (type `object`) and boolean (type `bool`) features you can use the **`value_counts`** method. Let's look at the target feature (`y`) distribution:


In [12]:
df["y"].value_counts()

no     36548
yes     4640
Name: y, dtype: int64

4640 clients (11.3%) of 41188 issued a term deposit, the value of the variable `y` equals `yes`.

Let's look at the client distribution by the variable `marital`. We specify the value of the `normalize = True` parameter to view relative frequencies, but not absolute.

In [13]:
df["marital"].value_counts(normalize = True)

married    0.61
single     0.28
divorced   0.11
unknown    0.00
Name: marital, dtype: float64

As we can see, 61% (0.61) of clients are married, which must be taken into account when planning marketing campaigns to manage deposit operations.


### Sorting

A `DataFrame` can be sorted by a few feature values. In our case, for example, by `duration` (`ascending = False` for sorting in descending order):

In [14]:
df.sort_values(by = "duration", ascending = False).head()

age          job  marital            education default housing loan  \
24091   33   technician   single  professional.course      no     yes   no   
22192   52  blue-collar  married             basic.4y      no      no   no   
40537   27       admin.   single          high.school      no      no   no   
13820   31   technician  married  professional.course      no      no   no   
7727    37   unemployed  married  professional.course      no     yes   no   

         contact month day_of_week  ...  campaign  pdays  previous  \
24091  telephone   nov         mon  ...         1    999         0   
22192  telephone   aug         thu  ...         3    999         0   
40537  telephone   aug         fri  ...         1    999         0   
13820   cellular   jul         thu  ...         1    999         0   
7727   telephone   may         fri  ...         2    999         0   

          poutcome emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
24091  nonexistent        -0.10           93.20         -42.00       4.41   
22192  nonexistent         1.40           93.44         -36.10       4.96   
40537  nonexistent        -1.70           94.03         -38.30       0.89   
13820  nonexistent         1.40           93.92         -42.70       4.96   
7727   nonexistent         1.10           93.99         -36.40       4.86   

       nr.employed    y  
24091      5195.80   no  
22192      5228.10  yes  
40537      4991.60   no  
13820      5228.10  yes  
7727       5191.00  yes  

[5 rows x 21 columns]

The sorting results show that the longest calls exceed one hour, as the value `duration` is more than 3600 seconds or 1 hour. At the same time, it usually was on Mondays and Thursdays (`day_of_week`) and, especially, in November and August (`month`).


Sort by the column group:


In [15]:
df.sort_values(by = ["age", "duration"], ascending = [True, False]).head()

age      job marital education default  housing     loan   contact  \
38274   17  student  single   unknown      no       no      yes  cellular   
37579   17  student  single  basic.9y      no  unknown  unknown  cellular   
37140   17  student  single   unknown      no      yes       no  cellular   
37539   17  student  single  basic.9y      no      yes       no  cellular   
37558   17  student  single  basic.9y      no      yes       no  cellular   

      month day_of_week  ...  campaign  pdays  previous  poutcome  \
38274   oct         tue  ...         1      2         2   success   
37579   aug         fri  ...         2    999         1   failure   
37140   aug         wed  ...         3      4         2   success   
37539   aug         fri  ...         2    999         2   failure   
37558   aug         fri  ...         3      4         2   success   

      emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  nr.employed    y  
38274        -3.40           92.43         -26.90       0.74      5017.50  yes  
37579        -2.90           92.20         -31.40       0.87      5076.20  yes  
37140        -2.90           92.20         -31.40       0.88      5076.20   no  
37539        -2.90           92.20         -31.40       0.87      5076.20   no  
37558        -2.90           92.20         -31.40       0.87      5076.20   no  

[5 rows x 21 columns]

We see that the youngest customers are at the `age` of 17, and the call `duration` exceeded 3 minutes only for three clients, which indicates the ineffectiveness of long-term interaction with such clients.


### Application of functions: `apply`, `map` etc.

**Apply the function to each column:**


In [16]:
df.apply(np.max)

age                      98
job                 unknown
marital             unknown
education           unknown
default                 yes
housing                 yes
loan                    yes
contact           telephone
month                   sep
day_of_week             wed
duration               4918
campaign                 56
pdays                   999
previous                  7
poutcome            success
emp.var.rate           1.40
cons.price.idx        94.77
cons.conf.idx        -26.90
euribor3m              5.04
nr.employed         5228.10
y                       yes
dtype: object

The oldest client is 98 years old (`age` = 98), and the number of contacts with one of the customers reached 56 (`campaign` = 56).

The `apply` method can also be used to apply the function to each row. To do this, you need to specify the `axis = 1`.

**Apply the function to each column cell**

The `map` can also be used for **the values ​​replacement in a column** by passing it as an argument dictionary in form of ` {old_value: new_value}  `.


In [17]:
d = {"no": 0, "yes": 1}
df["y"] = df["y"].map(d)
df.head()

age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

  month day_of_week  ...  campaign  pdays  previous     poutcome emp.var.rate  \
0   may         mon  ...         1    999         0  nonexistent         1.10   
1   may         mon  ...         1    999         0  nonexistent         1.10   
2   may         mon  ...         1    999         0  nonexistent         1.10   
3   may         mon  ...         1    999         0  nonexistent         1.10   
4   may         mon  ...         1    999         0  nonexistent         1.10   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed  y  
0           93.99         -36.40       4.86      5191.00  0  
1           93.99         -36.40       4.86      5191.00  0  
2           93.99         -36.40       4.86      5191.00  0  
3           93.99         -36.40       4.86      5191.00  0  
4           93.99         -36.40       4.86      5191.00  0  

[5 rows x 21 columns]

### Indexing and extracting data


A `DataFrame` can be indexed in many ways. In this regard, consider various ways of indexing and extracting data from the DataFrame with simple question examples.

You can use the code `dataframe ['name']` to extract a separate column. We use this to answer the question: **What is the share of clients attracted in our DataFrame?**

In [18]:
print("Share of attracted clients =", '{:.1%}'.format(df["y"].mean()))

Share of attracted clients = 11.3%


11,3% is a rather bad indicator for a bank, with such a percentage of attracted customers a business can collapse.

Logical indexation by one column of a `DataFrame` is very convenient. It looks like this: `df [p(df['Name']]`, where`  p ` is a certain logical condition that is checked for each element of the `Name` column. The result of such an indexation is a `DataFrame` consisting only of the rows satisfying the condition `p` by the `Name` column.

We use this to answer the question: **What are the mean values ​​of numerical features among the attracted clients?**

In [19]:
df[df["y"] == 1].mean()

age                40.91
duration          553.19
campaign            2.05
pdays             792.04
previous            0.49
emp.var.rate       -1.23
cons.price.idx     93.35
cons.conf.idx     -39.79
euribor3m           2.12
nr.employed      5095.12
y                   1.00
dtype: float64

Thus, the average age of the attracted clients is about 40 (`age` = 40.91), and 2 calls were required to attract them (`campaign` = 2.05).

Combining two previous types of indexation, we will answer the question: **What is the average call duration for the attracted clients**?


In [20]:
acd = round(df[df["y"] == 1]["duration"].mean(), 2)
acd_in_min = acd // 60
print("Average call duration for attracted clients =", acd_in_min, "min", int(acd) % 60, "sec")

Average call duration for attracted clients = 9.0 min 13 sec


So, the average duration of a successful call is almost 553 seconds, that is, nearly 10 minutes.

**What is the average age of attracted (`y == 1`) and unmarried (`'marital' == 'single'`) clients?**



In [21]:
print("Average age of attracted clients =", int(df[(df["y"] == 1) & (df["marital"] == "single")]["age"].mean()), "years")

Average age of attracted clients = 31 years
